components/brave_wallet/browser/meld_integration_service.cc

```diff
@@ -0,0 +1,549 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/meld_integration_service.h"
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/containers/flat_map.h"
+#include "base/environment.h"
+#include "base/logging.h"
+#include "base/strings/stringprintf.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/browser/json_rpc_requests_helper.h"
+#include "brave/components/brave_wallet/browser/meld_integration_response_parser.h"
+#include "brave/components/brave_wallet/common/buildflags.h"
+#include "brave/components/json/rs/src/lib.rs.h"
+#include "net/base/url_util.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace {
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("meld_integration_service", R"(
+      semantics {
+        sender: "Meld Integration Service"
+        description:
+          "This service is used to obtain assets prices from the external
+ Meld API  for the Brave wallet."
+        trigger:
+          "Triggered by uses of the native Brave wallet."
+        data:
+          "Meld JSON RPC response bodies."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature on chrome://flags."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+base::flat_map<std::string, std::string> MakeMeldApiHeaders() {
+  base::flat_map<std::string, std::string> request_headers;
+  std::unique_ptr<base::Environment> env(base::Environment::Create());
+  std::string meld_api_key(BUILDFLAG(MELD_API_KEY));
+  if (env->HasVar("MELD_API_KEY")) {
+    env->GetVar("MELD_API_KEY", &meld_api_key);
+  }
+  meld_api_key = base::StrCat({"BASIC", " ", meld_api_key});
+  request_headers["Authorization"] = std::move(meld_api_key);
+  request_headers["accept"] = "application/json";
+  request_headers[brave_wallet::kMeldRpcVersionHeader] =
+      brave_wallet::kMeldRpcVersion;
+
+  return request_headers;
+}
+constexpr char kDefaultMeldStatuses[] = "LIVE,RECENTLY_ADDED";
+
+std::optional<std::string> SanitizeJson(const std::string& json) {
```

Comment:
What is the maximum size of `json`?
If it's large enough, two calls via RustCxx make a copy of the string multiple times.
Could we just make one call something like `convert_all_numbers_to_string_and_remove_null_value` to reduce the overhead?

Reply:
> What is the maximum size of json?

The maximum length depends from filtering parameters, according to my observation, if there are no any filtering applied, the length of json is ~360 lines

> Could we just make one call something like convert_all_numbers_to_string_and_remove_null_value to reduce the overhead?

I think it makes sense, I'll change it

Reply:
It looks like running this sanitization as a conversion callback would be sanitizing an untrusted input before we're able to [safely sanitize it with the JSON parser](https://github.com/brave/brave-core/blob/219f661a74c713ee9c1cb15cd68a6fcd708c29b6/components/api_request_helper/api_request_helper.cc#L475-L478) that adhere's to the [Rule of 2](https://chromium.googlesource.com/chromium/src/+/master/docs/security/rule-of-2.md).

I've opened a slack thread on this because it seems to be a problem more broad then just this PR: https://bravesoftware.slack.com/archives/CU5UXM3EC/p1715130869577699

Reply:
verified my understanding on this wasn't correct. This is being handled in Rust so adheres to the Rule of 2.